### PR TITLE
Change datatypes of TripDay, TripTime, and Price

### LINKED ISSUE #99: Refactor TripDay, TripTime and Price to use native datatypes

### PR DESCRIPTION
**Pull request by ZechariahTan**

### CODE REVIEW

File: src/main/java/seedu/address/model/person/passenger/Passenger.java [hunk 1]
```diff
@@ -75,10 +75,18 @@ public TripDay getTripDay() {
         return tripDay;
     }
 
+    public String getTripDayStr() {
```
Comment (chewterence) [inline src/main/java/seedu/address/model/person/passenger/Passenger.java]: small thing but, perhaps this could be renamed getTripDayAsString() as we have discussed

File: src/main/java/seedu/address/model/person/passenger/Passenger.java [hunk 2]
```diff
@@ -75,10 +75,18 @@ public TripDay getTripDay() {
...
+        return tripDay.toString();
+    }
+
     public TripTime getTripTime() {
         return tripTime;
     }
 
+    public String getTripTimeStr() {
```
Comment (chewterence) [inline src/main/java/seedu/address/model/person/passenger/Passenger.java]: small thing but, perhaps this could be renamed getTripTimeAsString() as we have discussed

File: src/main/java/seedu/address/storage/JsonAdaptedPassenger.java [hunk 3]
```diff
@@ -65,9 +70,10 @@ public JsonAdaptedPassenger(Passenger source) {
         name = source.getName().fullName;
         phone = source.getPhone().value;
         address = source.getAddress().value;
-        tripDay = source.getTripDay().value;
-        tripTime = source.getTripTime().value;
-        driver = source.getDriverStr();
+        tripDayStr = source.getTripDay().toString();
+        tripTimeStr = source.getTripTime().toString();
```
Comment (chewterence) [inline src/main/java/seedu/address/storage/JsonAdaptedPassenger.java]: perhaps we should call TripDay and TripTime's source.getTripDayAsString() method so as to not violate the law of demeter?

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 4]
```diff
@@ -36,14 +38,22 @@
...
-    public static final String VALID_PRICE_AMY = "1.69";
-    public static final String VALID_PRICE_BOB = "6.9";
+    public static final DayOfWeek VALID_TRIPDAY_FRIDAY = DayOfWeek.FRIDAY;
+    public static final DayOfWeek VALID_TRIPDAY_MONDAY = DayOfWeek.MONDAY;
+    public static final LocalTime VALID_TRIPTIME_EVENING = LocalTime.of(18, 00);
+    public static final LocalTime VALID_TRIPTIME_MORNING = LocalTime.of(8, 30);
+    public static final DayOfWeek VALID_TRIPDAY_BOB = DayOfWeek.THURSDAY;
+    public static final LocalTime VALID_TRIPTIME_BOB = LocalTime.of(19, 30);
+    public static final double VALID_PRICE_AMY = 1.69;
+    public static final double VALID_PRICE_BOB = 6.9;
```
Comment (chewterence) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java:48]: should these be the actual values in the native types?
or should these be the strings, because i believe command test util is to test the input that the user puts in (in the form of a string). To test if it is correct or wrong input

File: src/test/java/seedu/address/model/person/passenger/PriceTest.java [hunk 5]
```diff
@@ -8,14 +8,9 @@
...
 
-    @Test
-    public void constructor_null_throwsNullPointerException() {
-        assertThrows(NullPointerException.class, () -> new Price(null));
-    }
-
     @Test
     public void constructor_invalidPrice_throwsIllegalArgumentException() {
-        String invalidPrice = "";
+        double invalidPrice = 12000.09310;
```
Comment (chewterence) [inline src/test/java/seedu/address/model/person/passenger/PriceTest.java]: this is ok because its testing the constructor, not the command

File: src/test/java/seedu/address/model/person/passenger/TripDayTest.java [hunk 6]
```diff
@@ -13,34 +14,12 @@ public void constructor_null_throwsNullPointerException() {
...
-    }
-
     @Test
     public void isEqual() {
-        TripDay mon = new TripDay("monday");
-        TripDay mon2 = new TripDay("MONDAY");
+        DayOfWeek day1 = DayOfWeek.valueOf("MONDAY");
+        TripDay mon = new TripDay(day1);
+        DayOfWeek day2 = DayOfWeek.valueOf("MONDAY");
+        TripDay mon2 = new TripDay(day2);
```
Comment (chewterence) [inline src/test/java/seedu/address/model/person/passenger/TripDayTest.java:22]: this is ok because its testing the .equals method, not the command

File: src/test/java/seedu/address/testutil/PassengerBuilder.java [hunk 7]
```diff
@@ -23,9 +25,9 @@
     public static final String DEFAULT_NAME = "Amy Bee";
     public static final String DEFAULT_PHONE = "85355255";
     public static final String DEFAULT_ADDRESS = "123, Jurong West Ave 6, #08-111";
-    public static final String DEFAULT_TRIPDAY = "FRIDAY";
-    public static final String DEFAULT_TRIPTIME = "1800";
-    public static final String DEFAULT_PRICE = "1.69";
+    public static final DayOfWeek DEFAULT_TRIPDAY = DayOfWeek.FRIDAY;
+    public static final LocalTime DEFAULT_TRIPTIME = LocalTime.of(18, 0);
```
Comment (chewterence) [inline src/test/java/seedu/address/testutil/PassengerBuilder.java:29]: i believe this should be ok, because its for the purpose of constructing a passenger

**Review by chewterence — COMMENTED** [review]
method renaming could be done in another PR, other than that do take a look at my comments 👍

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java (hunk 4 above)
Comment (ZechariahTan) [reply src/test/java/seedu/address/logic/commands/CommandTestUtil.java:48]: Thanks for the comment! To clarify, the VALID_[VARIABLE] constants are not used as replacements for user input but rather for the personbuilder and other such utility functions during testing. As for tests concerning user inputs, the constants declared below of the form [VARIABLE]_DESC_BOB, and as they are supposed to emulate user inputs, they are of the String datatype. Hope this addresses your comment!

File: src/main/java/seedu/address/storage/JsonAdaptedPassenger.java (hunk 3 above)
Comment (ZechariahTan) [reply src/main/java/seedu/address/storage/JsonAdaptedPassenger.java]: Thanks for the comment! I will make the relevant changes.

File: src/main/java/seedu/address/model/person/passenger/Price.java [hunk 8]
```diff
@@ -13,16 +12,15 @@
     public static final String MESSAGE_CONSTRAINTS =
             "The price you are willing to pay should be entered as a positive decimal number.";
     public static final String VALIDATION_REGEX = "\\d+(\\.\\d\\d?)?";
```
Comment (markmcwong) [inline src/main/java/seedu/address/model/person/passenger/Price.java:15]: If I'm not wrong, the regex doesn't validate whether the number given is a negative number or not right?
maybe can do `^[1-9]` at start of the pattern to make sure there's no negative sign I guess?

File: src/test/java/seedu/address/model/person/passenger/PriceTest.java (hunk 5 above)
Comment (markmcwong) [inline src/test/java/seedu/address/model/person/passenger/PriceTest.java]: Can also test for negative price as those should be invalid too!

**Review by markmcwong — COMMENTED** [review]
LGTM, maybe you can filter out the negative double out from user also

File: src/main/java/seedu/address/model/person/passenger/Passenger.java [hunk 9]
```diff
@@ -75,19 +75,27 @@ public TripDay getTripDay() {
...
+    public String getTripDayAsStr() {
+        return tripDay.toString();
+    }
+
     public TripTime getTripTime() {
         return tripTime;
     }
 
+    public String getTripTimeAsStr() {
+        return tripTime.toString();
```
Comment (lyueyang) [inline src/main/java/seedu/address/model/person/passenger/Passenger.java:87]: is this necessary? thinking because you could just use TripTime's internal toString() method implicitly but it could also be that im not too familiar with this

File: src/main/java/seedu/address/model/person/passenger/Price.java [hunk 10]
```diff
@@ -13,16 +12,15 @@
...
 
     /**
      * Constructs a {@code Price}.
      *
      * @param price A valid price.
      */
-    public Price(String price) {
-        requireNonNull(price);
-        checkArgument(isValidPrice(price), MESSAGE_CONSTRAINTS);
+    public Price(double price) {
```
Comment (lyueyang) [inline src/main/java/seedu/address/model/person/passenger/Price.java:23]: maybe keeping the requireNonNull might be useful since that's what we're encouraged to do this week

**Review by lyueyang — APPROVED** [review]
just some small tweaks that you could consider making, not necessary, just for you to consider, beyond that LGTM!

File: src/main/java/seedu/address/model/person/passenger/Price.java (hunk 8 above)
Comment (ZechariahTan) [reply src/main/java/seedu/address/model/person/passenger/Price.java:15]: Oh nice catch! I'll make the required changes!

File: src/main/java/seedu/address/model/person/passenger/Passenger.java (hunk 9 above)
Comment (ZechariahTan) [reply src/main/java/seedu/address/model/person/passenger/Passenger.java:87]: Thanks for bringing this up. I think it might be useful as it would prevent the violation of the Law of Demeter when other classes try to call for the TripTime strings.

File: src/main/java/seedu/address/model/person/passenger/Price.java (hunk 10 above)
Comment (ZechariahTan) [reply src/main/java/seedu/address/model/person/passenger/Price.java:23]: Good point. I removed it as Intellij suggested doing so as double data types aren't nullable but I'll add it in just in case.

File: src/main/java/seedu/address/model/person/passenger/Price.java (hunk 8 above)
Comment (ZechariahTan) [reply src/main/java/seedu/address/model/person/passenger/Price.java:15]: Nice catch! I didn't think of it but I realised when testing that the \d+ already catches negative cases as "-" is not a digit and so the input string would fail the isValidPrice check!
